### PR TITLE
Improve ProcessStatusResponse

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Improve ProcessStatusResponse signature so that we can tell if the
error is caused by malformed status response or the error is from stored
value inside status response. 
This PR is cut from https://github.com/project-chip/connectedhomeip/pull/19356.

#### Change overview
See above

#### Testing
no logic change, the existing test covers.

### CODE REVIEW

File: src/app/CommandSender.cpp [hunk 1]
```diff
@@ -130,7 +130,10 @@ CHIP_ERROR CommandSender::OnMessageReceived(Messaging::ExchangeContext * apExcha
 
     if (mState == State::AwaitingTimedStatus)
     {
-        err = HandleTimedStatus(aPayloadHeader, std::move(aPayload));
+        CHIP_ERROR statusError = CHIP_NO_ERROR;
+        SuccessOrExit(err = StatusResponse::ProcessStatusResponse(std::move(aPayload), statusError));
```
Comment (bzbarsky-apple) [inline src/app/CommandSender.cpp:134]: This needs to first check that the message is in fact a StatusResponse, no?  This is why there was a helper method for this not-entirely-trivial logic in `TimedRequest`....

File: src/app/WriteClient.cpp [hunk 2]
```diff
@@ -431,7 +431,11 @@ CHIP_ERROR WriteClient::OnMessageReceived(Messaging::ExchangeContext * apExchang
 
     if (mState == State::AwaitingTimedStatus)
     {
-        err = HandleTimedStatus(aPayloadHeader, std::move(aPayload));
+        CHIP_ERROR statusError = CHIP_NO_ERROR;
+        SuccessOrExit(err = StatusResponse::ProcessStatusResponse(std::move(aPayload), statusError));
```
Comment (bzbarsky-apple) [inline src/app/WriteClient.cpp:435]: This needs to first check that the message is in fact a StatusResponse.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/CommandSender.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/CommandSender.cpp:135]: yes, just add this check back. thanks for detailed check.

File: src/app/WriteClient.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/WriteClient.cpp:436]: yes, just add this check back. thanks for detailed check.

File: src/app/CommandSender.cpp [hunk 3]
```diff
@@ -130,6 +130,7 @@ CHIP_ERROR CommandSender::OnMessageReceived(Messaging::ExchangeContext * apExcha
 
     if (mState == State::AwaitingTimedStatus)
     {
+        VerifyOrReturnError(aPayloadHeader.HasMessageType(MsgType::StatusResponse), CHIP_ERROR_INVALID_MESSAGE_TYPE);
```
Comment (bzbarsky-apple) [inline src/app/CommandSender.cpp:133]: This needs to go to the exit label, not return?

File: src/app/WriteClient.cpp [hunk 4]
```diff
@@ -431,6 +431,7 @@ CHIP_ERROR WriteClient::OnMessageReceived(Messaging::ExchangeContext * apExchang
 
     if (mState == State::AwaitingTimedStatus)
     {
+        VerifyOrReturnError(aPayloadHeader.HasMessageType(MsgType::StatusResponse), CHIP_ERROR_INVALID_MESSAGE_TYPE);
```
Comment (bzbarsky-apple) [inline src/app/WriteClient.cpp:434]: Needs to go to the exit label.

File: src/app/CommandSender.cpp (hunk 3 above)
Comment (bzbarsky-apple) [reply src/app/CommandSender.cpp:133]: And set err, for that matter.  So probably:
```
  VerifyOrExit(aPayloadHeader.HasMessageType(MsgType::StatusResponse), err = CHIP_ERROR_INVALID_MESSAGE_TYPE);
```

File: src/app/CommandSender.cpp [hunk 5]
```diff
@@ -130,7 +130,11 @@ CHIP_ERROR CommandSender::OnMessageReceived(Messaging::ExchangeContext * apExcha
 
     if (mState == State::AwaitingTimedStatus)
     {
-        err = HandleTimedStatus(aPayloadHeader, std::move(aPayload));
+        VerifyOrReturnError(aPayloadHeader.HasMessageType(MsgType::StatusResponse), CHIP_ERROR_INVALID_MESSAGE_TYPE);
```
Comment (bzbarsky-apple) [inline src/app/CommandSender.cpp:133]:
```suggestion
        VerifyOrExit(aPayloadHeader.HasMessageType(MsgType::StatusResponse), err = CHIP_ERROR_INVALID_MESSAGE_TYPE);
```

File: src/app/WriteClient.cpp [hunk 6]
```diff
@@ -431,7 +431,12 @@ CHIP_ERROR WriteClient::OnMessageReceived(Messaging::ExchangeContext * apExchang
 
     if (mState == State::AwaitingTimedStatus)
     {
-        err = HandleTimedStatus(aPayloadHeader, std::move(aPayload));
+        VerifyOrReturnError(aPayloadHeader.HasMessageType(MsgType::StatusResponse), CHIP_ERROR_INVALID_MESSAGE_TYPE);
```
Comment (bzbarsky-apple) [inline src/app/WriteClient.cpp:434]:
```suggestion
        VerifyOrExitaPayloadHeader.HasMessageType(MsgType::StatusResponse), err = CHIP_ERROR_INVALID_MESSAGE_TYPE);
```

**Review by bzbarsky-apple — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]

**Pull request merged**